### PR TITLE
PRO-2338

### PR DESCRIPTION
**Pull request by boutell**
Smaller PR today:

* Reset expectations some more in the docs.
* Moved the automatic inlining of certain `require` files to an `--upgrade-required-files` command line option, which will be documented as probably too pre-alpha for most users, at least for now.
* Fixed a crash when the upgrader auto-renames a piece module.
* The linter can now flag module names ending in -widgets, spot lib/modules, and detect project level folders for legacy core modules which need to be renamed.
* The linter engine now supports custom functions for both matching and output purposes.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,27 +10,38 @@
...
-The Code Upgrader handles the majority of required modifications for an Apostrophe 2 (A2) codebase to run Apostrophe 3 (A3). It will also identify many specific lines and sections of code that a developer will need to convert manually.
+The Code Upgrader handles a portion of the required modifications for an Apostrophe 2 (A2) codebase to run Apostrophe 3 (A3). It will also identify many specific lines and sections of code that a developer will need to convert manually.
 
 **Status:** In development (not for production use)
 
 ## Purpose
 
 ### What it does
 
+#### The `lint` command
```
Comment (abea) [inline README.md]: This is meant to be an overview of the tool as a whole. The lint command material should be at the start of  the "Linting the A2 codebase" section.

File: README.md [hunk 2]
```diff
@@ -10,27 +10,38 @@
...
+
+The `lint` command scans your project for modifications that likely need to be made to be compatible with A3. The lint command will work well with basically all projects and detects many issues. Here are just a few examples of what the linter can detect:
+
+- The need to rename `lib/modules` to `modules`.
+- The need to change `{{ apos.area(...) }}` to `{% area ... %}`.
+- The need to move code from `construct()` to `methods()`, `handlers()`, etc.
+
+Since the linter is very tolerant it is a good candidate for use with nearly all A2 projects that are migrating to A3.
+
+#### The `upgrade` command
```
Comment (abea) [inline README.md]: Similarly, this material should be at the start of the "Running the code upgrade" section below.

File: README.md [hunk 3]
```diff
@@ -10,27 +10,38 @@
...
-- Converting field schemas, columns, and similarly structured features to the A3 "cascade" configuration structure.
-- Converting utility methods such as `addHelper()`, `apiRoute()`, and others to A3 module customization functions.
-- Moving code in `beforeConstruct`, `construct`, and `afterConstruct` that can't otherwise be converted into appropriate A3 module functions.
-- Inlining the content of any files used in module configuration with the `require('./lib/something')(self, options)` pattern. This allows the tool to run the rest of its actions on that code as well. Those complete "required" files are removed.
+- Converting field schemas, columns, and similarly structured features to the A3 "cascade" configuration structure, if the existing module follows the structure of our A2 example projects closely enough.
+- Converting utility methods such as `addHelper()`, `apiRoute()`, and others to A3 module customization functions, again if the project closely follows the structure of our typical projects.
+- Moving code in `construct`, and `afterConstruct` that can't otherwise be converted into appropriate A3 module functions.
```
Comment (abea) [inline README.md]: I did see `beforeConstruct` material moved to `beforeSuperClass`. has that changed?

File: README.md [hunk 4]
```diff
@@ -65,12 +77,16 @@ Use the command `apos-code-upgrader lint` to run a linter scan of the A2 code. T
...
 3. Review the changes (before committing them) with `git status` and `git diff HEAD`. Even though files are moved and directories renamed, git will still be able to display line changes for most of them.
 
 ### Reset the changes before committing
 
-If you want to undo all the changes made by the tool for any reason, run `apos-code-upgrader reset` in the project root. You must do this *before committing the changes*. This does a simple hard reset with git and will let you start again after reviewing the changes if you desire.
+If you want to undo all the changes made by the tool for any reason, run `apos-code-upgrader reset` in the project root. You must do this *before committing the changes*. This does a simple hard reset of your local branch with git and will let you start again after reviewing the changes if you desire.
 
 As a last resort you can always switch back to the main git branch and create a new upgrade branch to start over. You did switch off the main branch at the start, didn't you? 🤓
+
+## Pre-alpha options
```
Comment (abea) [inline README.md]: This should be a third level heading before the reset bit.

File: README.md [hunk 5]
```diff
@@ -65,12 +77,16 @@ Use the command `apos-code-upgrader lint` to run a linter scan of the A2 code. T
...
+
+The upgrade command has experimental support for inlining certain `require`-d files in order to discover methods, handlers, etc. inside those files as well. This generally will not work well unless required files confine their interesting logic to an exported function, which `index.js` invokes with `(self, options)`. You can try out this experimental feature by adding the `--upgrade-required-files` flag.
```
Comment (abea) [inline README.md]: For clarity I suggest:

> ... support for inlining certain module configuration included with `require` statements in order to discover...

Comment (abea) [reply README.md]: "confine" => "limit"

File: README.md (hunk 4 above)
Comment (abea) [reply README.md]: I don't think we need to specify that they are pre-alpha. Simply "Options." As it evolves out of alpha the documentation will change.

**Review by abea — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -10,48 +10,66 @@
...
+
+- The need to rename `lib/modules` to `modules`.
+- The need to change `{{ apos.area(...) }}` to `{% area ... %}`.
+- The need to move code from `construct()` to `methods()`, `handlers()`, etc.
+
+Since the linter is very tolerant it is a good candidate for use with nearly all A2 projects that are migrating to A3.
+
+#### Automated upgrading
+
+This module can also carry out some upgrades automatically. While this is a great feature, keep in mind that not every module and project is a good candidate for automated code upgrading as there are an infinite number of ways projects can be structured. The upgrade featureworks best on projects that adhere very closely to the coding style of the official A2 sample projects and documentation.
```
Comment (abea) [inline README.md]: Typo "featureworks"

File: README.md [hunk 7]
```diff
@@ -10,48 +10,66 @@
...
+
+There are also many needed changes that the upgrade command cannot handle on its own. So a successful upgrade will always involve reviewing the output of the linting feature,  as well as the Apostrophe documentation.
```
Comment (abea) [inline README.md]: two spaces after "the linting feature,"

File: README.md [hunk 8]
```diff
@@ -10,48 +10,66 @@
...
 ### What it does
 
-The code upgrader will convert Apostrophe 2 codebases for installable modules *and* full A2 websites so they are *mostly ready* to run Apostrophe 3. This includes:
+This module's features break down into two basic categories: "linting" for compatibility issues you can fix yourself, and "upgrading" code automatically. Since not everything can be upgraded automatically, the linting feature is important for everyone.
+
+#### Linting for compatibility issues
+
+This module's linting feature scans your project for modifications that likely need to be made to be compatible with A3. The lint command will work well with basically all projects and detects many issues. Here are just a few examples of what the linter can detect:
+
+modifications that likely need to be made to be compatible with A3. The lint command will work well with basically all projects and detects many issues. Here are just a few examples of what the linter can detect:
```
Comment (abea) [inline README.md]: This is duplicated from above.

File: README.md [hunk 9]
```diff
@@ -10,48 +10,66 @@
...
+
+There are also many needed changes that the upgrade command cannot handle on its own. So a successful upgrade will always involve reviewing the output of the linting feature,  as well as the Apostrophe documentation.
+
+Where possible, the code upgrader will convert Apostrophe 2 codebases for installable modules *and* full A2 websites so they are *mostly ready* to run Apostrophe 3. This includes:
 - Moving modules from `lib/modules` to the `modules` directory.
 - Renaming most project-level Apostrophe core module customization directories to the A3 equivalents.
-- Converting field schemas, columns, and similarly structured features to the A3 "cascade" configuration structure.
-- Converting utility methods such as `addHelper()`, `apiRoute()`, and others to A3 module customization functions.
+- Converting field schemas, columns, and similarly structured features to the A3 "cascade" configuration structure, if the existing module follows the structure of our A2 example projects closely enough.
```
Comment (abea) [inline README.md]: I think we should avoid "our." In the original I'd worded it as "the official A2 sample projects" to not speak from a place of ownership.

File: README.md [hunk 10]
```diff
@@ -63,14 +81,25 @@ The upgrader tool will change most files in the codebase. It is important to pre
 
 Use the command `apos-code-upgrader lint` to run a linter scan of the A2 code. This will print to the console every required change it can find. This will *not* actually change any code. It is an especially useful step after you run the upgrade process, but it can be useful before to understand what changes to expect.
 
-## Running the code upgrade
+## Automatically upgarding the A2 codebase
```
Comment (abea) [inline README.md]: "upgarding"

**Review by abea — CHANGES_REQUESTED** [review]
a few bits to clean up

File: README.md (hunk 9 above)
Comment (abea) [reply README.md]: You got the one below. Another on L42.

**Review by abea — CHANGES_REQUESTED** [review]
one missed text thing. Also the ESLint errs

**Comment by boutell** [comment]
I think you said there is still one text duplication issue somehow but I can't find it?

**Comment by boutell** [comment]
(Line numbers have changed)

**Comment by abea** [comment]
@boutell It's the other "our A2 example projects"

**Review by abea — APPROVED** [review]

**Pull request merged**